### PR TITLE
Remove exp-verbs from public API

### LINKED ISSUE #79: get rid of _exp_  types in core.h
There is no good reason to require the user to use verbs exp types, which are not even available upstream.
Therefore, let's move to regular verbs types.

### PR DESCRIPTION
**Pull request by pakmarkthub**
**This change:**
- addresses issue #79.
- moves exp-verbs-related data structure from public API to internal API.
- enables libgdsync to be compatible with ib-verbs (non-exp) from the perspective of users.
- does not remove exp-verbs from the internal implementation of libgdsync.
- fixes issue #83.
- aligns code in all files.

**Presubmit Testings:**
- locally on two nodes. each node had:
  - GPU: P40 or P100
  - NVIDIA driver version: 418
  - CUDA version: 10.1
  - CPU: Intel(R) Xeon(R) CPU E5-2687W v4
  - NIC: Mellanox Technologies MT27700 Family [ConnectX-4]
  - OFED: MLNX_OFED_SRC-4.5-1.0.1.0
  - OS: Ubuntu 18.04.1 LTS
- Tests: gds_kernel_latency, gds_kernel_loopback_latency

### CODE REVIEW

File: src/gdsync.cpp [hunk 1]
```diff
@@ -1822,13 +1826,18 @@ struct gds_qp *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context,
                 return NULL;
         }
 
-        gqp = (struct gds_qp*)calloc(1, sizeof(struct gds_qp));
-        if (!gqp) {
+        memset(&exp_qp_attr, 0, sizeof(struct ibv_exp_qp_init_attr));
+        memcpy(&exp_qp_attr, qp_attr, sizeof(gds_qp_init_attr_t));
```
Comment (drossetti) [inline src/gdsync.cpp]: can you really bit copy ibv_qp_init_attr onto the ibv_exp_qp_init_attr version? is that guaranteed to work ?
@haggaie can you take a look?

File: include/gdsync/core.h [hunk 2]
```diff
@@ -54,8 +54,8 @@ enum gds_create_qp_flags {
     GDS_CREATE_QP_WQ_DBREC_ON_GPU = 1<<5,
 };
 
-typedef struct ibv_exp_qp_init_attr gds_qp_init_attr_t;
-typedef struct ibv_exp_send_wr gds_send_wr;
+typedef struct ibv_qp_init_attr gds_qp_init_attr_t;
+typedef struct ibv_send_wr gds_send_wr;
 
 struct gds_cq {
```
Comment (drossetti) [inline include/gdsync/core.h]: while we are at it, can we have typedefs for both gds_cq and gds_qp ?

File: src/gdsync.cpp [hunk 3]
```diff
@@ -1894,6 +1903,8 @@ struct gds_qp *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context,
         gqp->recv_cq.cq = qp->recv_cq;
         gqp->recv_cq.curr_offset = 0;
 
+        memcpy(qp_attr, &exp_qp_attr, sizeof(gds_qp_init_attr_t));
```
Comment (drossetti) [inline src/gdsync.cpp]: why is this needed?

File: include/gdsync/core.h [hunk 4]
```diff
@@ -166,10 +161,10 @@ int gds_stream_post_send_all(CUstream stream, int count, gds_send_request_t *req
  * Represents a wait operation on a particular CQ
  */
 
-typedef struct gds_wait_request {
-        struct ibv_exp_peer_peek peek;
-        struct peer_op_wr wr[GDS_WAIT_INFO_MAX_OPS];
-} gds_wait_request_t;
+typedef struct gds_wait_request gds_wait_request_t;
+
+int gds_alloc_wait_request(gds_wait_request_t **request, int num);
```
Comment (drossetti) [inline include/gdsync/core.h]: I'd rather not introduce ctors for request objects.
The usual trick is to define for example:
```
   typedef struct gds_wait_request *gds_wait_request_t;
```
And then to automatically allocate the objects internally to the APIs.

**Comment by drossetti** [comment]
As a reminder, this change will change the ABI.
So major version in master branch must be incremented before merging this one.

File: tests/gds_kernel_latency.c [hunk 5]
```diff
@@ -665,39 +709,44 @@ static int pp_post_work(struct pingpong_context *ctx, int n_posts, int rcnt, uin
...
                         if (ctx->use_desc_apis) {
-                                work_desc_t *wdesc = calloc(1, sizeof(*wdesc));
+                                work_desc_t *wdesc;
+                                if (alloc_work_desc(&wdesc) != 0) {
+                                    gpu_err("[%d] cannot alloc work desc\n", my_rank);
+                                    exit(EXIT_FAILURE);
+                                }
+
                                 int k = 0;
-                                ret = pp_prepare_gpu_send(ctx, qpn, &wdesc->send_rq);
```
Comment (drossetti) [inline tests/gds_kernel_latency.c]: if defined correctly, this should not be necessary

File: src/gdsync.cpp (hunk 1 above)
Comment (pakmarkthub) [reply src/gdsync.cpp]: This is the definition of struct ibv_qp_attr_attr and its relevant fields:

struct ibv_qp_cap {
    uint32_t        max_send_wr;
    uint32_t        max_recv_wr;
    uint32_t        max_send_sge;
    uint32_t        max_recv_sge;
    uint32_t        max_inline_data;
};

struct ibv_qp_init_attr {
    void               *qp_context;
    struct ibv_cq          *send_cq;
    struct ibv_cq          *recv_cq;
    struct ibv_srq         *srq;
    struct ibv_qp_cap   cap;
    enum ibv_qp_type    qp_type;
    int         sq_sig_all;
    /* Below is needed for backwards compatabile */
    struct ibv_xrc_domain  *xrc_domain;
};

struct ibv_exp_qp_init_attr has one-to-one mapping but with additional fields.

We mainly memcpy the pointers and some shallow structs. In other words, no deep copy seems to be needed. 

Note: We need to use ibv_exp_create_qp to support peer ops. Hence, directly passing ibv_qp_init_attr is not an option.

File: src/gdsync.cpp (hunk 3 above)
Comment (pakmarkthub) [reply src/gdsync.cpp]: The original algorithm sets some fields such as send_cq and recv_cq of qp_attr. Hence, the users can do the following:

gds_create_qp(..., &qp_attr, ...);
qp_attr.send_cq; // access send_cq here

This memcpy is for addressing this case.

File: include/gdsync/core.h (hunk 4 above)
Comment (pakmarkthub) [reply include/gdsync/core.h]: Some people recommend against hiding a pointer type behind typedef. One reason is that it can break expectation of "const". See https://wiki.sei.cmu.edu/confluence/display/c/DCL05-C.+Use+typedefs+of+non-pointer+types+only for more details.

Concerning ctors, what will you recommend instead?

File: src/gdsync.cpp (hunk 1 above)
Comment (haggaie) [reply src/gdsync.cpp]: In general the methodology of extending verbs would mean that it would be okay to memcpy, but it seems like in this case there's some incompatibility. The ibv_qp_init_attr of MLNX OFED includes the xrc_domain pointer for some reason where the comp_mask field would be in the exp version. I think the upstream version doesn't.

Comment (pakmarkthub) [reply src/gdsync.cpp]: I can copy field by field and leave out xrc_domain. However, if there is a change in the struct definition such as one more field is introduced, we may need to revisit this part of libgdsync again.

@haggaie Can you suggest a way to do this correctly such that it can withstand future changes in the struct definition?

Comment (haggaie) [reply src/gdsync.cpp]: Normally when extending these structs people try to preserve both ABI and source backward compatibility. Also, I don't think ibv_qp_init_attr can grow since it doesn't have a length field or a comp_mask field that would tell the callee which version of the struct it got. So I think it would be fine to copy field by field, or to use memcpy without the xrc_domain field.

**Comment by pakmarkthub** [comment]
@drossetti I addressed the above concerns with a new revision. Please re-review the code.

File: src/gdsync.cpp (hunk 1 above)
Comment (drossetti) [reply src/gdsync.cpp]: @haggaie thanks

File: src/gdsync.cpp (hunk 3 above)
Comment (drossetti) [reply src/gdsync.cpp]: Just to confirm, we don't need that, right? those fields are accessible thru gds_qp

File: include/gdsync/core.h [hunk 6]
```diff
@@ -279,8 +275,8 @@ typedef enum gds_tag {
 typedef struct gds_descriptor {
         gds_tag_t tag; /**< selector for union below */
         union {
-                gds_send_request_t  *send;
```
Comment (drossetti) [inline include/gdsync/core.h]: this API change and the next are not needed

File: include/gdsync/core.h (hunk 4 above)
Comment (drossetti) [reply include/gdsync/core.h]: I still don't think we should have explicit ctor/dtor APIs.
Are they really needed?

Comment (pakmarkthub) [reply include/gdsync/core.h]: No, we don't. I forgot to remove these functions. Will fix in the next revision.

File: src/gdsync.cpp (hunk 3 above)
Comment (pakmarkthub) [reply src/gdsync.cpp]: I will remove the copy in the next revision. Note that the behavior of gds_create_qp() will be changed because qp_attr will change from an in/out to in-only param.

**Comment by pakmarkthub** [comment]
Addressed all new comments.

File: src/utils.hpp [hunk 7]
```diff
@@ -193,9 +195,10 @@ typedef enum gds_alloc_qp_flags {
 
 typedef std::vector<CUstreamBatchMemOpParams> gds_op_list_t;
 
-struct gds_cq *gds_create_cq(struct ibv_context *context, int cqe, void *cq_context, struct ibv_comp_channel *channel, int comp_vector, int gpu_id, gds_alloc_cq_flags_t flags);
+gds_cq_t *gds_create_cq(struct ibv_context *context, int cqe, void *cq_context, struct ibv_comp_channel *channel, int comp_vector, int gpu_id, gds_alloc_cq_flags_t flags);
 int gds_post_pokes(CUstream stream, int count, gds_send_request_t *info, uint32_t *dw, uint32_t val);
-int gds_post_pokes_on_cpu(int count, gds_send_request_t *info, uint32_t *dw, uint32_t val);
+int gds_post_pokes_on_cpu(int count, gds_send_request_s *info, uint32_t *dw, uint32_t val);
+int gds_stream_post_wait_cq_multi(CUstream stream, int count, gds_wait_request_s *request, uint32_t *dw, uint32_t val);
```
Comment (drossetti) [inline src/utils.hpp]: are there really two variants of this function?

File: tests/gds_kernel_latency.c [hunk 8]
```diff
@@ -123,7 +124,7 @@ struct pingpong_context {
 	struct ibv_cq		*tx_cq;
 	struct ibv_cq		*rx_cq;
 	struct ibv_qp		*qp;
-	struct gds_qp		*gds_qp;
+	gds_qp_t		    *gds_qp;
```
Comment (drossetti) [inline tests/gds_kernel_latency.c]: alignment issue

File: tests/gds_kernel_latency.c [hunk 9]
```diff
@@ -542,16 +544,15 @@ static int pp_post_gpu_send(struct pingpong_context *ctx, uint32_t qpn, CUstream
 		.wr_id	    = PINGPONG_SEND_WRID,
 		.sg_list    = &list,
 		.num_sge    = 1,
-		.exp_opcode = IBV_EXP_WR_SEND,
-		.exp_send_flags = IBV_EXP_SEND_SIGNALED,
+		.opcode = IBV_WR_SEND,
```
Comment (drossetti) [inline tests/gds_kernel_latency.c]: alignment problem

**Comment by pakmarkthub** [comment]
Let's focus on fixing the issue here. The code style violation will be taken care of in another issue.

File: src/gdsync.cpp [hunk 10]
```diff
@@ -1241,37 +1243,44 @@ void gds_dump_wait_request(gds_wait_request_t *request, size_t count)
...
 
-        retcode=gds_stream_post_descriptors(stream, count, descs, 0);
-        if (retcode) {
-                gds_err("error %d in gds_stream_post_descriptors\n", retcode);
-                goto out;
+out:
+    if (descs) 
+        free(descs);
+    for (k = 0; k < count; ++k) {
+        if (request[k].handle != NULL && !(to_gds_wait_request_s(&request[k])->flags & GDS_FLAG_INTERNAL_KEEP_REQUESTS)) {
```
Comment (drossetti) [inline src/gdsync.cpp]: at least if calloc fails, I'm not sure you want to free user requests, as she/he might free memory and retry.
if gds_stream_post_descriptors(), we can argue.
it is probably a good idea to only free requests if ret==0

File: tests/gds_kernel_latency.c [hunk 11]
```diff
@@ -123,7 +124,7 @@ struct pingpong_context {
 	struct ibv_cq		*tx_cq;
 	struct ibv_cq		*rx_cq;
 	struct ibv_qp		*qp;
-	struct gds_qp		*gds_qp;
+    gds_qp_t            *gds_qp;
```
Comment (drossetti) [inline tests/gds_kernel_latency.c]: can you align this one?

**Comment by drossetti** [comment]
@pakmarkthub let's try to finalize this PR. Is anything else left to do?

Though we want to tag a release before merging this one. Can you do that?

**Comment by pakmarkthub** [comment]
@drossetti 

I draft a tag v2.2 here: https://github.com/gpudirect/libgdsync/releases/tag/untagged-5a92f02b250db5900964. Let me know if you want to mark it as release or not.

About this PR, I have aligned code in all files. Please check if it looks ok now.

File: include/gdsync/device.cuh [hunk 12]
```diff
@@ -33,129 +33,129 @@
...
-        T *ptr;
-        T value;
-
-        __host__ __device__ inline volatile T *access_once() {
-            return (volatile T *)ptr;
-        }
-        __host__ __device__ isem64() : ptr(NULL), value(0) {}
-    };
-    typedef struct isem64 isem64_t;
+        static const clock_t large_timeout = 1ULL<<32;
```
Comment (drossetti) [inline include/gdsync/device.cuh]: all these whitespace changes add a lot of noise, are they really needed?

File: src/apis.cpp [hunk 13]
```diff
@@ -90,18 +80,15 @@ static void gds_init_wait_request(gds_wait_request_t *request, uint32_t offset)
...
         int ret=0;
 
         assert(qp);
         assert(qp->qp);
         assert(send_info);
-        if(
-                        flag != IBV_EXP_ROLLBACK_ABORT_UNCOMMITED && 
-                        flag != IBV_EXP_ROLLBACK_ABORT_LATE
-          )
+        if (flag != IBV_EXP_ROLLBACK_ABORT_UNCOMMITED && flag != IBV_EXP_ROLLBACK_ABORT_LATE)
```
Comment (drossetti) [inline src/apis.cpp]: looks like this function hasn't been cleaned up

File: src/apis.cpp [hunk 14]
```diff
@@ -125,20 +112,26 @@ static int gds_rollback_qp(struct gds_qp *qp, gds_send_request_t * send_info, en
...
         }
 
-        ret = gds_post_pokes_on_cpu(1, &send_info, NULL, 0);
+        request = to_gds_send_request_s(&send_info);
+
+        ret = gds_post_pokes_on_cpu(1, request, NULL, 0);
         if (ret) {
                 gds_err("error %d in gds_post_pokes_on_cpu\n", ret);
-                ret_roll = gds_rollback_qp(qp, &send_info, IBV_EXP_ROLLBACK_ABORT_LATE);
+                ret_roll = gds_rollback_qp(qp, request, IBV_EXP_ROLLBACK_ABORT_LATE);
```
Comment (drossetti) [inline src/apis.cpp]: cleanup needed here as well

File: src/apis.cpp [hunk 15]
```diff
@@ -253,43 +263,51 @@ int gds_stream_post_send_all(CUstream stream, int count, gds_send_request_t *req
...
+        wait_request = (gds_wait_request_s *)malloc(sizeof(gds_wait_request_s));
+        if (wait_request == NULL) {
+                gds_err("out of memory in gds_prepare_wait_cq\n");
+                return ENOMEM;
+        }
+
+        gds_init_wait_request(wait_request, cq->curr_offset++);
 
-        retcode = ibv_exp_peer_peek_cq(cq->cq, &request->peek);
+        retcode = ibv_exp_peer_peek_cq(cq->cq, &wait_request->peek);
```
Comment (drossetti) [inline src/apis.cpp]: this is still an exp verb API call

File: src/gdsync.cpp [hunk 16]
```diff
@@ -1784,30 +1787,32 @@ gds_create_cq(struct ibv_context *context, int cqe,
...
+gds_qp_t *gds_create_qp(struct ibv_pd *pd, struct ibv_context *context,
+                gds_qp_init_attr_t *qp_attr, int gpu_id, int flags)
 {
         int ret = 0;
-        struct gds_qp *gqp = NULL;
+        gds_qp_internal_t *gqpi = NULL;
+        gds_qp_t *gqp = NULL;
         struct ibv_qp *qp = NULL;
-        struct gds_cq *rx_gcq = NULL, *tx_gcq = NULL;
+        struct ibv_exp_qp_init_attr exp_qp_attr;
```
Comment (drossetti) [inline src/gdsync.cpp]: exp verbs again

File: src/objs.hpp [hunk 17]
```diff
@@ -117,6 +120,64 @@ static inline gds_peer *peer_from_id(uint64_t id)
         return reinterpret_cast<gds_peer *>(id);
 }
 
+typedef struct {
+        gds_qp_t                   gqp;
+        struct ibv_exp_res_domain  *res_domain;
```
Comment (drossetti) [inline src/objs.hpp]: exp verbs usage?

File: src/objs.hpp [hunk 18]
```diff
@@ -117,6 +120,64 @@ static inline gds_peer *peer_from_id(uint64_t id)
...
+} gds_qp_internal_t;
+
+enum {
+        GDS_SEND_INFO_MAX_OPS = 32,
+        GDS_WAIT_INFO_MAX_OPS = 32
+};
+
+enum gds_flag_internal {
```
Comment (drossetti) [inline src/objs.hpp]: can we call this enum gds_request_flags ?

**Comment by pakmarkthub** [comment]
Better implementation in PR #86.

Closing this PR without merge.
